src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs

```diff
@@ -124,8 +124,22 @@ internal override IInvisibleEditor OpenInvisibleEditor(IVisualStudioHostDocument
         {
             // We need to ensure the file is saved, only if a global undo transaction is open
             var globalUndoService = this.Services.GetService<IGlobalUndoService>();
-            bool needsSave = globalUndoService.IsGlobalTransactionOpen(this);
-            bool needsUndoDisabled = needsSave && this.Services.GetService<IGeneratedCodeRecognitionService>().IsGeneratedCode(this.CurrentSolution.GetDocument(hostDocument.Id));
+            var needsSave = globalUndoService.IsGlobalTransactionOpen(this);
+
+            var needsUndoDisabled = false;
+            if (needsSave)
+            {
+                if (this.CurrentSolution.ContainsDocument(hostDocument.Id))
+                {
+                    // Disable undo on generated documents
+                    needsUndoDisabled = this.Services.GetService<IGeneratedCodeRecognitionService>().IsGeneratedCode(this.CurrentSolution.GetDocument(hostDocument.Id));
+                }
+                else
+                {
+                    // Enable undo on "additional documents" or if no document can be found.
+                    needsUndoDisabled = false;
```

Comment:
Yes. I could remove the ```else``` and keep the clarity with a different comment, but this felt like the most readable approach.